Neos.ContentRepository.Core/Classes/Feature/NodeCreation/Command/CreateNodeAggregateWithNodeAndSerializedProperties.php

```diff
@@ -104,12 +102,26 @@ public static function fromArray(array $array): self
         );
     }
 
+    public function withInitialPropertyValues(SerializedPropertyValues $newInitialPropertyValues): self
```

Comment:
Why was this added? initialPropertyValues can be set in the constructor already!?

Reply:
It was an inconsistency here its just to align the behaviour fully to `CreateNodeAggregateWithNode`. This serialised command is internal but i thought why not.

Reply:
Let me rephrase my question:
Why do we need a wither for parameters that are already part of the constructor?
I have the same question re `CreateNodeAggregateWithNode`. I traced it back to some special requirement in the [`NodeCreationCommands` of `Neos.UI`](https://github.com/neos/neos-ui/blob/bdff700ead29cd27b40f362a2beac84e9b19f63e/Classes/Domain/NodeCreation/NodeCreationCommands.php#L105).

IMO it's not the best idea to let those downstream dependencies control the architecture of the API (instead it should just collect all details before creating the command instances). But obviously this is not a big issue at all and it might even be useful for other, similar. requirements

Reply:
Kinda true i see. `CreateNodeAggregateWithNode::withInitialPropertyValues` probably always just existed for this particular case.

Annnnnd - trommel wirbel bitte 🥁 - you introduced it *g

https://github.com/neos/contentrepository-development-collection/commit/a73c8c768bbc471696d524bc57402e0bf97ced5a

Anyway with my neos ui adjustments and the [new NodeCreationCommands of Neos.UI](https://github.com/neos/neos-ui/blob/bdff700ead29cd27b40f362a2beac84e9b19f63e/Classes/Domain/NodeCreation/NodeCreationCommands.php#L105) we can actually get rid of this `with*er` and do it differently youre right.

So should i do it. Would be up for it.

Reply:
But no, that is super impractical actually. Because not everything can be set by the main constructor, we would have to write this:

```php
$newFirst = CreateNodeAggregateWithNode::create(
    $this->first->workspaceName,
    $this->first->nodeAggregateId,
    $this->first->nodeTypeName,
    $this->first->originDimensionSpacePoint,
    $this->first->parentNodeAggregateId,
    $this->first->succeedingSiblingNodeAggregateId,
    $newInitialPropertyValues
);

if ($this->first->nodeName !== null) {
    $newFirst = $newFirst->withNodeName($this->first->nodeName);
}

if ($this->first->tetheredDescendantNodeAggregateIds !== null) {
    $newFirst = $newFirst->withTetheredDescendantNodeAggregateIds($this->first->tetheredDescendantNodeAggregateIds);
}
```

would still be fine as its abstracted away in the Neos.Ui for the users, but we are creating so many intermediate instances and there is a possibility of bugs if you forget as user to call any `with*er` to transfer state.

idk.

Reply:
> Annnnnd - trommel wirbel bitte 🥁 - you introduced it *g

I can be - trommel wirbel bitte 🥁 - wrong

Reply:
But as written above, I think it's not an issue. I just wondered why you had added it in _this case_.

What we _could_ consider (for a separate PR): whether we want to create more generic withers for those cases. i.e:
```php
    private function with(
        public WorkspaceName $workspaceName = null,
        public NodeAggregateId $nodeAggregateId = null,
        public NodeTypeName $nodeTypeName = null,
        public OriginDimensionSpacePoint $originDimensionSpacePoint = null,
        public NodeAggregateId $parentNodeAggregateId = null,
        public SerializedPropertyValues $initialPropertyValues = null,
        public NodeAggregateId $succeedingSiblingNodeAggregateId = null,
        public NodeAggregateIdsByNodePaths $tetheredDescendantNodeAggregateIds = null,
    ) {
        return new self(
            $workspaceName ?? $this->workspaceName,
            $nodeAggregateId ?? $this->nodeAggregateId,
            $nodeTypeName ?? $this->nodeTypeName,
            $originDimensionSpacePoint ?? $this->originDimensionSpacePoint,
            $parentNodeAggregateId ?? $this->parentNodeAggregateId,
            $initialPropertyValues ?? $this->initialPropertyValues,
            $succeedingSiblingNodeAggregateId ?? $this->succeedingSiblingNodeAggregateId,
            $this->nodeName,
            $tetheredDescendantNodeAggregateIds ?? $this->tetheredDescendantNodeAggregateIds,
        );
    }
```

We did so in other places and IMO it works out nicely.
And in the future we might be able to replace the redundant implementation with [some clone](https://wiki.php.net/rfc/clone_with) with syntax

Reply:
> But as written above, I think it's not an issue. I just wondered why you had added it in _this_ case.

just for consistency. So _if_ someone is using this low level stuff on purpose it should not be too hard to use, but youre right its YAGNI - i also doubted if i should add it and rolled a dice :D

Reply:
Removed it ;)